### ISSUE TITLE
Implement quantile normalization

### ISSUE THREAD

**Issue by jaclyn-taroni** [AlexsLemonade/refinebio#488]
### Context

We'll be combining multiple platforms and even multiple technologies in this project, particularly in the species-wide compendium case. We'll want the underlying distributions of these platforms/technologies to be as similar as possible. 

We intend to use [quantile normalization](https://en.wikipedia.org/wiki/Quantile_normalization) to accomplish this.

By doing so, we assume, then, that the difference between samples arises solely from _technical differences_. Of course this is not always going to be the case. We'll use as many samples as possible to generate the reference or target distribution. By including as diverse biological conditions as we have available to us to inform the reference distribution, we hope that this will mitigate the effect of the "dampening" of biological signal that is likely to occur.

### Problem or idea

As I see it, this has three components:

1. Identify the most popular microarray platform for _each species_ (that we can process)
2. Calculate the reference distribution from the entire "most popular platform" compendium for each species
3. Just in time quantile normalization during smashing

_I'll note that this may be better split into multiple tickets for tracking, but we can make that determination once the team takes a look at it._

#### Most popular platform

We'll use microarray platforms because we, and others, have experience using microarray data as a reference distribution for RNA-seq.

Using the GEO data from #220, [`config/supported_microarray_platforms.csv`](https://github.com/AlexsLemonade/refinebio/blob/13444092d7c5f186612dd8228f3e656b42e3f895/config/supported_microarray_platforms.csv), and only including platforms with `num_samples > 100`, I get this list: [affymetrix_platforms_for_QN_reference.txt](https://github.com/AlexsLemonade/refinebio/files/2258673/affymetrix_platforms_for_QN_reference.txt)

We'll probably want to prioritize crunching samples on these platforms.

#### Calculate reference distribution

We should aggregate (smash) all samples together from a platform _that we have processed_ (e.g., do not include submitter-processed data) without any row-wise transformation.

We have some options here:

* With rpy2, we can use [`normalize.quantiles.determine.target`](https://www.rdocumentation.org/packages/preprocessCore/versions/1.34.0/topics/normalize.quantiles.target) from the `preprocessCore` R package. A downside raised here is reading the whole matrix into memory.
* I believe @kurtwheeler has an implementation in Python handy that works on a matrix and that includes unit tests. Not sure if/how it handles missing values, but I don't believe this will be too much of a concern if using `SCAN.UPC` processed Affymetrix data.
* Note that we don't necessarily need to smash everything together, just that all samples have to be taken into account in the calculations AND that the inner join that happens during smashing is helpful. I'll let whomever is doing the implementing make that choice.

**This process will need to happen periodically** as data gets added to the upstream repositories. We might upload these to Zenodo and grab at Docker build time similar to what we do for [AlexsLemonade/identifier-refinery](https://github.com/AlexsLemonade/identifier-refinery).

#### QN during smashing

We'll need to normalize the samples to the reference distribution from step 2 just in time during smashing.

Because different platforms measure different (numbers of) genes, their distributions are necessarily different lengths. The implementation will need to account for this. I believe the [`normalize.quantiles.use.target`](https://www.rdocumentation.org/packages/preprocessCore/versions/1.34.0/topics/normalize.quantiles.target) function in the `preprocessCore` R package does. I tested this by using example data with different numbers of rows (features/genes) and then comparing the distributions with a [Kolmogorov–Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov–Smirnov_test) ([`ks.test`](https://www.rdocumentation.org/packages/dgof/versions/1.2/topics/ks.test)). This could be used in conjunction with `rpy2`; the `target` is the reference distribution (e.g., vector) from step 2. In the future, we'll want to go about this differently (see below). 

There are two ways to approach this and I would like to get @cgreene's opinion on this. The `()` denotes an optional step that depends on the context.

**One:**
```
(1. Convert to Ensembl gene id) 
2. Collapse to mean 
3. Quantile normalize 
4. Aggregate 
(5. Transform row-wise)
```

If performed this way, the QN occurs on a single sample, with ameliorates some of the memory concerns of going the `rpy2` and `preprocessCore` route.

**Two:**
```
(1. Convert to Ensembl gene id) 
2. Collapse to mean 
3. Aggregate 
4. Quantile normalize 
(5. Transform row-wise)
```

If we aggregate after QN and multiple different platforms are included in this aggregation as in option one, the genes that are not in every sample will get dropped and the underlying distributions will then likely be different between samples. Option two, then, prioritizes having the same distribution. Note that this would require all samples being loaded into memory.

This quantile normalization solution _replaces_ an earlier version of the cross-technology smashing (https://github.com/AlexsLemonade/refinebio/issues/186#issuecomment-395218310, #330, #364).

### New Issue Checklist

- [x] The title is short and descriptive.
- [x] You have explained the context that led you to write this issue.
- [x] You have reported a problem or idea.
- [x] You have proposed a solution or next step.

#### Things to watch out for/future thoughts

These are mostly notes to self that came up along the way. They will become their own, more detailed tickets for a future release (at which point I'll remove this section). 

* Filtering low-count genes in RNA-seq data -- if less than some number of `lengthScaledTPM` in all the RNA-seq samples, drop row/gene?
* QN _only_ for genes that were used to derive the initial quantiles and then interpolate in between. This avoids a case such as HGU133A and HGU133B samples getting the same values which is likely to be wrong.
* Use all Affymetrix data for a species to get the reference distribution (e.g., more samples!), but we'll likely want to tackle something like #144 before we do that.
* At some point, the number of RNA-seq samples will surpass the number of microarray samples on a single platform. We should test this out.

**Comment by Miserlou** [comment]
![img_20180802_103336719](https://user-images.githubusercontent.com/139987/43664082-8f8bc992-9739-11e8-8ffa-ceb2ba6ce53b.jpg)
From the whiteboard sesh.

**Comment by jaclyn-taroni** [comment]
Note that that's a little bit different than what I wrote -- we don't need to quantile normalize the reference compendium before we calculate the target quantiles unless we want to store the QN data, which I don't think we do.

**Comment by cgreene** [comment]
☝️ I agree with not normalizing until preparation for delivery (smashing?) b/c the quantiles _could_ change with periodic releases and we would not want to have to re-normalize things.

**Comment by Miserlou** [comment]
@jaclyn-taroni Can you prepare some data so that we're not blocked by the Zebrafish crunch on this? Even if it's dummy data just so that we aren't blocked on the rest of the QN stuff by the need to scrape and process an entire organism.

**Comment by jaclyn-taroni** [comment]
Generated like so in R:
```R
set.seed(123)
mat1 <- matrix(rnorm(n = 300, mean = 0, sd = 1), ncol = 15)
mat2 <- matrix(rnorm(n = 300, mean = 3, sd = 5), ncol = 10)
write.table(mat1, "mat1.txt", quote = FALSE, sep = "\t")
write.table(mat2, "mat2.txt", quote = FALSE, sep = "\t")
```

[toy_data_for_488.zip](https://github.com/AlexsLemonade/refinebio/files/2267113/toy_data_for_488.zip)

This is random normally distributed data with different mean (`mean`) and standard deviation (`sd`) values. I've intentionally generated these such that they have different numbers of rows/genes/features.

If you would like to increase the number of rows, increase the `n` argument to `rnorm`. If you'd like additional columns/samples, increase the `ncol` argument to `matrix`.

**Comment by jaclyn-taroni** [comment]
@cgreene I am still interested in your thoughts about the order of steps during the smashing/preparation for delivery process: option one vs. option two in **QN during smashing**. (I believe your comment above was about storing the QN reference data, but I may be mistaken.)

**Comment by cgreene** [comment]
I don't understand the difference between option one vs. option two. If you are using target quantiles, would the produced output be effectively the same? I would imagine that each row gets mapped to those quantiles.

**Comment by jaclyn-taroni** [comment]
I guess my point re: option one is depending on which genes/how many get dropped in the normalized data during the aggregation step (because we drop genes that are not present in every sample), your distribution could end up looking a bit different between different platforms. I'm not sure that this is a big deal in practice and I also lean towards this option.

**Comment by jaclyn-taroni** [comment]
Option two it is 👍

**Comment by jaclyn-taroni** [comment]
Some thoughts that have come up while discussing this + looking at Jackiecrunch data:

* We should store/keep track of the genes that were included in the compendium used to generate the reference distribution. We likely won't use this information _yet_, but will be good for the future.
* Affymetrix data that has been processed with SCAN + Ensembl gene id Brainarray appears to, for the human samples I've looked at, retain control probes (e.g., `AFFX-`). I think we should probably drop those.
